### ISSUE TITLE
Support connecting to multiple networks

### ISSUE THREAD

**Issue by kylrth** [tailscale/tailscale#183]
We use Tailscale at work, and I'd like to use Tailscale at home. As such I want to be able to connect to the two networks simultaneously. When I was using `relay-node`, I created two separate systemd services with separate `--tun`s and `--port`s. When I switched to `tailscaled`, I had to work through all of that again, updating my service files to look like the default one.

It seems like my solution is probably not the best one. Is there a way that a single `tailscaled` process could manage the two networks?

**Comment by danderson** [comment]
This is something we plan to support (I've been calling it "personas" to distinguish it from the OAuth accounts and the local user accounts, both of which are also involved in some ways with tailscaled), and we also feel this pain right now (we use Tailscale at Tailscale Inc., but that makes it hard to set up a test network to try out something that'd break prod).

No timeline for when it'd happen yet, but - yes, we feel your pain, and we want to fix it :)

**Comment by danderson** [comment]
I should add: for now, your "run two tailscaled's" hack should work, albeit not perfectly - they'll be fighting over netfilter rules a little bit, but not in a harmful way right now. There's some upcoming network stack changes that I'd have to think through that might break stuff (basically, any time one tailscaled restarts the other will experience a short outage). We'll try to keep this working, even though it's not a configuration we want to support long-term.

**Comment by kylrth** [comment]
Yeah I've had some issues with it for sure. Keep me updated.

**Comment by tylerb** [comment]
Adding my +1 here as well. I've just started using Tailscale, and I can see a need for at least 3 separate simultaneous logins.

**Comment by JayDoubleu** [comment]
@danderson Would this also fall under category of "enable multi account support" ? was meaning to create an issue for that.

**Comment by kylrth** [comment]
Does anyone have a semi-functioning hack for this? I had to remove my home network service entirely, because the very fact that I had two made both unstable, even if only one was running.

**Comment by JayDoubleu** [comment]
@kylrth multipass might be a good workaround meanwhile.  https://github.com/canonical/multipass

**Comment by kylrth** [comment]
My hack with systemd broke with a recent update, so I'm just using a single "persona" for now (using SOCKS proxy to a device with the other persona to access devices on that network). Thought I'd bump this issue.

**Comment by coffenbacher** [comment]
I'm sure you guys know how important this one is, but to share a quick anecdote: 

My company is working on a team project with two S&P 500 companies and Tailscale would be a perfect fit for parts. But we already use Tailscale, so we can't juggle our internal network with another one.  I had to stop my evangelizing mid-sentence when I realized that blocker.

**Comment by mattapperson** [comment]
Any updates on official support for this or working workarounds?

**Comment by Plasma** [comment]
This would be great to support connecting to different environments as well.

**Comment by GlidLov** [comment]
+1 on this

**Comment by ben-z** [comment]
@GlidLov please use reactions instead of commenting and notifying everyone who's subscribed to this issue.

**Comment by JayDoubleu** [comment]
I don't mind being notified personally. It's good to remind this is indeed desired by customers.

**Comment by pmarreck** [comment]
I only recently discovered Tailscale, been playing with it, and think it's **awesome**.
But being able to be a member of multiple Tailscale networks would be **N times** as awesome.
Thing is, even if you figure out a DNS scheme (such as subdomaining nodes off a network name to know to which network to steer lookups to), all the networks would have to agree on who's using what IP address range in order not to collide/overlap (so for example they couldn't both just use 100.* addresses as you'd have IP address collision pretty quickly... what if you were connected to 10 networks and some of those people were connected to 10 OTHER networks, etc.? And who's then enforcing global name uniqueness, Tailscale I guess?), and at that point you're basically re-solving the original Internet DNS/ipv4 problems or doing NAT between all these networks all over again, except at another layer (!).
It suddenly seems significantly harder than just implementing 1 network at a time per device. Classic scaling problem where going from 0-1 is an order of magnitude easier than going from 1-N.
I wonder if ipv6 or MAC addresses might actually help here with the routing/uniqueness problem?

**Comment by yuvadm** [comment]
Simultaneously-connected networks are an advanced feature, IMO. Even having the ability to easily switch contexts between different networks would be super useful. This would be an ideal start for this issue without tackling the harder problems.

**Comment by vistarjs** [comment]
A lot of our team members use tailscale for their personal devices. Multiple networks would allow us to manage this without having the set up complex acls for their personal devices on our corporate tailnet.

**Comment by xdevs23** [comment]
Maybe another way could be to connect to tailnets together in the admin panel and use ACLs to grant cross-tailnet access for users/groups. Is this something that will be implemented?

**Comment by solarkraft** [comment]
Sorry for yet another +1 post, but this is a pretty big deal. Tailscale's pricing model suggests that you *want* technical people to first set up Tailscale networks at home, start loving it and then implement it in organisations they're affiliated with. This is a great tactic and it has worked on me, but the inability to simultaneously connect to multiple tailnets effectively penalizes this flow.

**Comment by bradfitz** [comment]
We know. This impacts everybody who works at Tailscale too. We all have personal accounts. We're going to do this. It's not trivial though and has many dependencies and some other tasks are more important.

**Comment by yuvadm** [comment]
@bradfitz does "simple" context switching between different networks (as described in https://github.com/tailscale/tailscale/issues/183#issuecomment-1018348651) also include non-trivial tasks?

**Comment by DentonGentry** [comment]
Seriously: we know people want this. We know. I promise we will update this bug when there is progress to report. We're not going to write up estimates of every variation.

**Comment by afdaniele** [comment]
@AndreaCensi I guess the world is a small place after all. I didn't know you were using it already when you suggested it to us (coworkers) this morning. :smile: 
I ended up on this issue because I too was looking for ways to connect the same machine to different tailnets. 

Keep up the good work guys, we'll wait.

**Comment by will** [comment]
I'd potentially like to connect a server to two completely different and unrelated tailscale networks, and have it accept inbound traffic from both. Not a desktop or laptop making outbound connections. I have no need for the server to make any outbound connections over tailscale.

Is this issue the one I should be watching or is what I'm looking for different than this issue?

**Comment by DentonGentry** [comment]
A potential solution which would work now is to start the server in a third tailnet, and then share it into the other two: https://tailscale.com/kb/1084/sharing/

Shared nodes can accept incoming connections from the tailnet they are shared into, but not make outgoing connections.

**Comment by bt** [comment]
Honestly, I'd be happy enough right now with a workaround for `Switch User`, but in a more streamlined way.

I swap between my personal and work tailnets so often that it's gotten annoying to have to re-log every time, and re-assign ACL tags.

**Comment by andygikling** [comment]
Can we get an update on this? @apenwarr?

Tailscale you're making real money these days - please work on this much needed feature!  

As @AndreaCensi pointed out - this is a logical show-stopper for mass adoption.  If you're painted yourself into a corner with MagicDNS, just invest the money and fix it please!

ZeroTier has this feature folks...

**Comment by apenwarr** [comment]
As others have mentioned, we're working on it. We're not satisfied with the security results from the most obvious solutions, which create a lot of inconvenience (user switching every time you want to ssh to a different device isn't as fun as it sounds) or security problems (one device connected to multiple networks creates lateral attack vectors between multiple networks).

Sorry for the inconvenience in the meantime. As others have mentioned, we're feeling it too.

**Comment by pmarreck** [comment]
> one device connected to multiple networks creates lateral attack vectors between multiple networks

hooking up 2 network cards to my computer which are connected to 2 different networks (I guess this would technically turn it into a gateway device, but really, any device that is connected to, say, both a wifi network and gigabit wired might comprise this) presents the same risks; is this really your problem to solve? Of course, I'm sure a network admin at one's office that uses Tailscale would want a say about their devices also being connected to other TS networks simultaneously (such as a users' homelab/LAN stuff), in fact, maybe there's a business use-case there waiting for you to grab, but to some extent (short of complete hardware lockdown, such as what my S.O. has to deal with at her job), this seems like a not-completely-preventable existential risk; the proper firewall and security setup (which you could document, potentially, or set as a default, such as disallowing cross-traffic between networks), and simply "trusting people," may go a long way here, no?

Basically, how is the security risk here any different from what has been available to be done "manually" for years, except that the networks are virtual instead of physical?

**Comment by solarkraft** [comment]
If my computer gets infected from one network, don't you think connecting to the other network at all (not at the same time) would also be a problem?

It'd be a nice quick workaround to be able to switch between multiple logged in accounts. I think that's 90% of what people here want: Some way to avoid having to log out and back in all the time.

**Comment by apenwarr** [comment]
I think you can see in the above two comments that the problem is rather complex and deserves attention. In general, at Tailscale we're willing to sacrifice a bit of usage growth and convenience in order to not create accidental security risks. The result is it does take us longer to take the necessary care to launch privacy/security-affecting features like this.

We're working on it!

**Comment by edude03** [comment]
I hate to be "that guy" who comes across as entitled and whiny on GitHub issues (and I hope it's not taken that way) but I do want to say that this issue and the related issue of [fast user switching](https://github.com/tailscale/tailscale/issues/183#issuecomment-598462418) has been open for 2.5 years now. I'm a generally happy tailscale user who has gotten a few of my clients to use tailscale (and pay for it) so it's extra frustrating whenever I just need to do a quick "kubectl get pods" and I have to go through a whole song and dance to switch accounts. 

It's gotten to the point that it's actually faster to just use two computers to do my job (and yes, I realized I'm very privileged) than it is to switch tailscale accounts multiple times throughout the day. 

I understand that you've heard us and you're working on it - but in case there is any doubt, "just" being able to switch accounts without having to sign in again would solve 90% of the problem with (presumably) much less work than solving the edge cases of multiple networks at the same time, so if at all possible, please give us a crumb and add fast user switching first.

**Comment by ocharles** [comment]
@edude03 What is your goal with this comment? It's been said /numerous/ times that asking "when will this be ready" will just get the response "when it's ready.". There's no industry where saying "I want this" gets things faster. You say you hate to be that guy, so why are you being that guy? I think TS should just lock this issue so we can remain subscribed for actual updates, and get rid of the monthly "polite" nudges.

**Comment by DentonGentry** [comment]
I've locked this issue, at least for a time.

**Comment by DentonGentry** [comment]
Tailscale 1.34 added Fast User Switching. While still logged into only one tailnet at a time, it does make it much easier to switch between them.
https://tailscale.com/blog/fast-user-switching/

**Comment by apollo13** [comment]
Once this feature is implemented will it allow one to only use multiple accounts at tailscale.com or would it also be possible to use one account at tailscale.com and one account against a private headscale instance. I am obviously hoping for the second option but would understand if you wouldn't want to support such a usecase.

**Comment by DentonGentry** [comment]
This feature is not currently being actively worked on by an engineering team, right now we do not have answers of what might be supported.

**Comment by pmarreck** [comment]
Fast user switching is a welcome feature.

Am I to surmise that the technical/practical issue with a device being on 2+ different tailnets at once is that it also then becomes a potential gateway between those tailnets which opens up quite the potential can of worms?

Fast user switching conveniently dodges this problem while also making working with multiple configured tailnets easier. Nicely done.

**Comment by mkaminsky** [comment]
In case it's useful to others, here's what I've been doing to connect to machines on two different tailnets, e.g., work and home.  Depending on your environment and application requirements, it might be a workable solution.

The basic idea is to disable kernel-model Tailscale and use multiple usermode instances instead.  Then, configure the applications that need TS to use the respective proxies.

### Setup

I'm using a Linux laptop.  I disable the default kernel-mode TS.  Then,

```bash
$ tailscaled --tun=userspace-networking --socks5-server=localhost:1055 --socket /home/kaminsky/.local/share/tailscale/tailscaled.sock --statedir /home/kaminsky/.local/share/tailscale/ &
$ tailscaled --tun=userspace-networking --socks5-server=localhost:1056 --socket /home/kaminsky/.local/share/tailscale2/tailscaled.sock --statedir /home/kaminsky/.local/share/tailscale2/ &
```

At this point, once signed-in, the following commands show the machines on each tailnet:
```bash
$ tailscale --socket /home/kaminsky/.local/share/tailscale/tailscaled.sock status
...

$ tailscale --socket /home/kaminsky/.local/share/tailscale2/tailscaled.sock status
...
```

### Applications

The main reason why this works for me, is that I primarily use only two applications/protocols over TS:  ssh and http(s).

Both are easy to configure to use the SOCKS proxy and can be set up to use the proxy based on machine names/patterns.

**ssh**

In `~/.ssh/config`, I have something like this:
```
Host *.tailABCDE.ts.net
    ProxyCommand=nc -x localhost:1056 %h %p

Host home-server
    HostName server.tailABCDE.ts.net
    ProxyCommand=nc -x localhost:1056 %h %p

Host work-server
    HostName server.tail12345.ts.net
    ProxyCommand=nc -x localhost:1055 %h %p
```

**https/browser**

Most/all browsers have proxy support, and they also have extensions/addons that make using multiple proxies easy.  For example, you can configure the proxy addons to send certain domains to proxy1 (port 1055) and others to proxy2 (port 1056).  Or, you could use different browser profiles, or entirely different browsers.  Lots of options here.

In both cases, DNS queries go to the respective proxies/tailscales and everything works pretty seamlessly.  

### Summary

The only thing to remember is to launch the respective userspace tailscaleds after a reboot, etc.  I wonder if adding some support/syntactic sugar to the tailscale CLI tool would be useful/make sense—just to avoid having to deal with the various socket params when trying to get a "status" dump.  Of course, shell aliases also work. :)

Tailscale is great--thanks to everyone involved!

**Comment by drio** [comment]
Also, I wrote this (in case you find it useful): https://gist.github.com/drio/b90b772a2857e873bf95214841ee95d1.

**Comment by cdhowie** [comment]
> The only thing to remember is to launch the respective userspace tailscaleds after a reboot, etc.

Creating a systemd unit for each instance should be pretty straightforward, I would think. Perhaps even one template service file could work for many instances without needing to create a `.service` for each.

**Comment by Seas0** [comment]
Making a instantized systemd unit file and spawning multiple instances of `tailscaled`, with setting `"netfilterMode": "off"` in all (including main) `tailscaled` configs or `tailscale up --netfilter-mode off` should allow connecting to multiple tailnet at the same time.
This trick **completely disabled** firewall rule creation, so it would definitely break features requiring kernel firewall framework, like Exit-Node, etc.. Since I'm not using Exit-Node or other features that necessitating firewall I'm not fixing it, but manually managing firewall rules should theoretically resume these features. Also, confliction of IPs is possible, so use with cautious.
A sample systemd unit and JSON config for `tailscaled` written by me:
```ini
[Unit]
Description=Tailscale node agent instance #%i
Documentation=https://tailscale.com/kb/
Wants=network-pre.target
After=network-pre.target NetworkManager.service systemd-resolved.service

[Service]
EnvironmentFile=/etc/tailscale.d/tailscale%i.env
ExecStart=/usr/sbin/tailscaled --config=/etc/tailscale.d/tailscale%i.json --state=/var/lib/tailscale%i/tailscaled.state --socket=/run/tailscale%i/tailscaled.sock --tun tailscale%i --port=${PORT} $FLAGS
ExecStopPost=/usr/sbin/tailscaled --cleanup --config=/etc/tailscale.d/tailscale%i.json --state=/var/lib/tailscale%i/tailscaled.state --socket=/run/tailscale%i/tailscaled.sock --tun tailscale%i

Restart=on-failure

RuntimeDirectory=tailscale%i
RuntimeDirectoryMode=0755
StateDirectory=tailscale%i
StateDirectoryMode=0700
CacheDirectory=tailscale%i
CacheDirectoryMode=0750
Type=notify

[Install]
WantedBy=multi-user.target
```
The declarative configuration file for `tailscaled` is still in its infancy without a thorough document and is subject to change in the future, but here is the current `alpha0` version configuration template extract from [ConfigVAlpha](https://github.com/tailscale/tailscale/blob/main/ipn/conf.go#L15)
```json
{
  "version": "alpha0",
  "locked": false,
  "serverURL": "https://controlplane.tailscale.com",
  "authKey": null,
  "enabled": true,
  "operatorUser": null,
  "hostname": null,
  "acceptDNS": true,
  "acceptRoutes": null,
  "allowLANWhileUsingExitNode": false,
  "advertiseRoutes": [],
  "disableSNAT": false,
  "netfilterMode": "off",
  "postureChecking": false,
  "runSSHServer": null,
  "runWebClient": null,
  "shieldsUp": false,
  "autoUpdate": {
    "check": true,
    "apply": false
  },
  "serveConfigTemp": null
}
```
Finally, the detailed steps for creating a new `tailscaled` instance
1. Create the `tailscaled@.service` unit file at `/etc/systemd/system/tailscaled@.service` with forementioned content
2. Make a folder called `/etc/tailscale.d`
3. Place configuration JSON file at `/etc/tailscale.d/tailscale1.json`
4. Copy the environment variable file from `/etc/default/tailscaled` to `/etc/tailscale.d/tailscale1.env`, change its content accordingly. (At least choose another listening port)
5. Enable and start `tailscaled` service instance 1 with `sudo systemctl enable --now tailscaled@1.service`

**Comment by novirium** [comment]
This might be a little naive, but does getting this working not just boil down to allowing the netfilter/iptables rules from multiple `tailscaled` instances to coexist? Something like adding an instance suffix to the chain names and the packet marking value? Almost everything else in `tailscaled` already allows instances to be separated (particularly if they're being run in containers).

For the subset of users (like me) that would like access to two tailnets on one device (particularly with subnet routing), I suspect setting up two instances manually would be a perfectly fine solution. Multiple tailnets doesn't have to be a complex feature added into `tailscaled` itself - it just needs to be able to avoid clobbering other instances of it (with some work by the user).

**Comment by rsyring** [comment]
> This might be a little naive, but does getting this working not just boil down to allowing the netfilter/iptables rules from multiple tailscaled instances to coexist?

I do think its not likely that easy.  From the note the end of the post just before yours where that is what they got working:

> DISCLAIMER: This trick completely disabled tailscale's automatic (iptables|nftables) firewall rule creation (or the rules created by tailscaled started earlier would be wiped by the latter one and lock you outside the former), so it would definitely break features requiring kernel firewall framework, like Exit-Node (with its dependency to a masqurade rule in NAT chains), MSS clamping to PMTU, etc.. Since I'm not using Exit-Node or other features that necessitating firewall, I'm not fixing it anyway, but in theory manually managed firewall rules should resume these features. Also, confliction of IPs is possible, and could possibly mess up the in-kernel IP routing table, so use with cautious.

A summary of thei ssues then could be:

1. IP overlap...that's a pretty significant one IMO.  Fixing it would probably require adding IP allocation ranges to subnets that were going to exist together on the same machine?  Just like you have to do with DHCP.  Maybe there is some other fancy trick to come up with but this seems like a fundamental problem to this working.
2. You'd likely want the tailscale client to manage both networks, not have two daemons running, to avoid conflicts mentioned in the firewall.  I can imagine that's a huge undertaking b/c the client's architecture was built with the assumption of managing one tailnet.  So potentially a complete rewrite or, at least, a complete re-architecture.
3. An employee mentioned concerns about security previously

I'm not affiliated with the project in any way, so could be off base here.  Just thought I'd try a shot at a summary for others who hit this issue and wonder why it can't be a simple solution.

**Comment by realyukii** [comment]
> Making a instantized systemd unit file and spawning multiple instances of tailscaled, with setting "netfilterMode": "off" in all (including main) tailscaled configs or with tailscale up --netfilter-mode off should allow connecting to multiple tailnet at the same time while not resort to pure userspace networking and keep Magic DNS integration with systemd-resolved.

when `tailscaled` is running in multiple instances, how it knows which account should be used/how it determine which tailnet should be connected?

currently my approach is switch the user with `sudo tailscale switch <registered-email-account@domain>` but it's not 'simultaneously' and I'm not sure how to setup multiple instances with given systemd unit examples

**Comment by Seas0** [comment]
> > Making a instantized systemd unit file and spawning multiple instances of tailscaled, with setting "netfilterMode": "off" in all (including main) tailscaled configs or with tailscale up --netfilter-mode off should allow connecting to multiple tailnet at the same time while not resort to pure userspace networking and keep Magic DNS integration with systemd-resolved.
> 
> when `tailscaled` is running in multiple instances, how it knows which account should be used/how it determine which tailnet should be connected?
> 
> currently my approach is switch the user with `sudo tailscale switch <registered-email-account@domain>` but it's not 'simultaneously' and I'm not sure how to setup multiple instances with given systemd unit examples

`tailscale` have a `--socket </path/to/tailscaled/socket>` commandline flag, you should use different controlling socket for different `tailscaled` instances

**Comment by realyukii** [comment]
on the new socket I can't switch to another account, it says: `No profile named "account@mail.com"`

is there a way to import the profile?

**Comment by Seas0** [comment]
> on the new socket I can't switch to another account, it says: `No profile named "account@mail.com"`
> 
> is there a way to import the profile?

Manually transfer the `profile-xxxx` entries in `/var/lib/tailscale/tailscaled.state` to corresponding state file of new `tailscaled` (mine is `/var/lib/tailscale1/tailscaled.state` should be suffice.

**Comment by slayer** [comment]
any official solutions? plans to support?

**Comment by 5jvm0u4** [comment]
Hi, same issue here from five years later, just want to join the waitlist.

**Comment by Dampfwalze** [comment]
>   1. IP overlap...that's a pretty significant one IMO.  Fixing it would probably require adding IP allocation ranges to subnets that were going to exist together on the same machine?  Just like you have to do with DHCP.  Maybe there is some other fancy trick to come up with but this seems like a fundamental problem to this working.

The least thing the client can do is to notify the user of an overlap.

A solid implementation would additionally let the user choose how to deal with the Situation. The user should have the following options:

1. Disable either of the conflicting routes (on this client), so that at least one of them will work. 
2. (Maybe) disable both routes.
3. Resolve the conflict manually by changing either nodes IP on the tailnet (in the admin console).

Until a user has manually chosen one over the other, the default behaviour should be either to disable both routes automatically, or leave only the route from the persona configured first on the client be enabled.

When a route is disabled, only the respective IP protocol should be disabled. If there is an IPv4 conflict, IPv6 should still work fine. MagicDNS should also be aware of disabled nodes and only resolve the respective protocols IP address that is not disabled.

For user-owned nodes, notifying could mean, getting something like a system notification (and/or maybe an email), and/or notify him while configuring a new persona. For a tag-owned node, the person to notify is usually an administrator. In that case, the notification could appear in the admin console, or as an email to an administrator in charge. Although connecting a server instance to multiple tailnets is arguably not a good practice anyway, but maybe sometimes it is not easily avoidable.

**Comment by mazunki** [comment]
https://tailscale.com/blog/multiple-tailnets-alpha

Edit:
Not sure why people are downvoting. I was subscribed to this post waiting for an update, and would have appreciated getting the notification that this feature is now in alpha before a friend actually told me about it.

**Comment by djc** [comment]
Because this issue is about connecting with multiple identities simultaneously, rather than splitting networks within a single identity provider.

**Comment by AetherMagee** [comment]
> https://tailscale.com/blog/multiple-tailnets-alpha

I don't think this is what is requested in this issue?

The request is for a *single client* to connect to *multiple, completely distinct* tailnets. Maybe one from the place they work at, and another to their homelab.

This blog describes multiple tailnets for a single org, which is... just ACLs on steroids I guess?

**Comment by mazunki** [comment]
I see. I might have misunderstood the blogpost: I implicitly assumed it would support connecting to several of those nets at once, like one can be member of multiple groups within an ACL, not only allow identity providers to create them.

In that case, I'm not sure what the blogpost actually is adding that policy tables don't already offer.

**Comment by erick-souz** [comment]
I'm facing the same problem and am wonder if Docker containers on a custom network could be a workaround solution. I'm not very experienced with networking, so I'm wondering if anyone has already tried this and what the results were.

**Comment by DXist** [comment]
As I understand tailnet daemon applies a single network policy defined by the active telnet - this is right from the network admin perspective.

If somebody wants to access services from another tailnet they should be shared with the current tailnet and approved by the network admins of both tailnets.

**Comment by madjam002** [comment]
> If somebody wants to access services from another tailnet they should be shared with the current tailnet and approved by the network admins of both tailnets.

Workarounds exist for this issue as is, plus people can just connect two different VPNs to bridge networks together if they really wanted to, so it's not really a reason for this feature to not be implemented.

Plus Tailscale has the concept of device posture checks, so if this feature were ever to be implemented, there could be a posture check for e.g `node:tsMultipleTailnets` where access to secure resources could be restricted if the client was connected to multiple tailnets, if you really needed to.

This feature would be extremely helpful for what seems like a common use case where you want to be connected to a personal tailnet and corporate tailnet at the same time. E.g as a small business owner, I am connected to my personal NAS at home for access to files, whilst simultaneously connected to work resources (SSHed into servers, access to shared test environments, etc). Also the use case of two companies wanting to connect to each other and both using Tailscale (as mentioned earlier in the thread).

In terms of IP overlap, I like the default of just disabling the routes until further action is taken. Same I think should happen with 4via6 subnets and just subnet routing in general, if there are overlapping routes, they should just be disabled. Perhaps also an option where the multiple tailnet connections are in a certain order and can be reordered, and then the routes take effect based on whichever one is first.

**Comment by Crank-Git** [comment]
I kind of overcame this issue inside of Linux ([hydrascale](https://github.com/Crank-Git/Hydrascale)). It runs each tailnet inside of its own Linux network namespace with a dedicated tailscaled instance, so traffice between the tailnets never meet. A veth pair connects namespaces back to the host and if you use the host_access option, hydrascale adds host routes for every peer's tailscale ip and manages /etc/hosts entries so you can just do things like 'ping corp-server' or 'ssh homelab-nas' without any wrappers. The host just sees all of the tailnet peers as a flat namespace with $tailnet-$hostname naming to avoid collisions.

Not a Tailscale solution, but a fun workaround that allows me to serve things to specific tailnets.

E1: Please feel free to use it, break it, and then add issues to fix
E2: I do understand that this is not an answer to the original ask "Is there a way that a single tailscaled process could manage the two networks?," just wanted to let people know there are other ways about it on Linux.

**Comment by Bnjoroge1** [comment]
My approach to solve this on a mac is to run a proxy server. I opt for system containers like Orbstack machines or Colima vs full VMs because they are much lighter. Works mostly well - just gotta remember to include proxyJump command in your ssh config. On linux, hydrascale above seems like a great option or you can build your own using namespaces like I describe in this [post](https://bnjoroge.com/posts/connecting-to-multiple-tailscale-networks-on-a-single-host/)

**Comment by Seas0** [comment]
Optionally, use [sing-box](https://github.com/SagerNet/sing-box) as a universal routing platform with properly configured Tailscale endpoints and DNS/routing paths. If all you need is multi-proxy access from a local workstation, this should cover it.
